Small fixes of issues with old devices support, brightness, etc

avently opened the pull request:
<!-- Hey there. Thank you so much for improving NewPipe. Please take a moment to fill out the following suggestion on how to structure this PR description. Having roughly the same layout helps everyone considerably :)-->

#### What is it?
- [x] Bug fix (user facing)

#### Description of the changes in your PR
Some very small issues were fixed by this PR. 
- I don't like this thing https://github.com/TeamNewPipe/NewPipe/compare/dev...avently:small-fixes2?expand=1#diff-20e0b2ce67b307277c153b396fb79660R355  but making the height anything but wrap content makes red dot in progress bar misaligned on Android 4.4, see #4245. With default height progress bar has smaller touchable area but looks ok on old devices too.
- this thing https://github.com/TeamNewPipe/NewPipe/compare/dev...avently:small-fixes2?expand=1#diff-57eb59a31e38fed8f965fb4a544e3b9fR540 prevents freeze on 4.4. Looks like some devices have even less available size for storing bundle data than Google recommends, see #4230. Anyway, this is how it should work and I'm fine with it
- brightness will be saved in the place where it changes.

#### Fixes the following issue(s)
closes #4271 
closes #4253 
closes #4245 
closes #4230 
closes #4277 
closes #4143
closes #4294 
closes #4292 
closes #4322

#### Testing apk
find the testing apk somewhere in this page because it's changes often

#### Agreement
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.

MD77MD commented:
brightness levels work ok now👌

opusforlife2 commented:
@avently The seekbar jumps up a pixel in this as well. To be clear, it's not just the seekbar that moves. The two timers and the fullscreen button also move with it. The whole bottom setup also shrinks by a pixel on either side, I just noticed.

avently commented:
@opusforlife2 it's probably happens because in the app status bar in landscape mode and have one size, then you move out of the app, the status bar in portrait and it's larger, when you move Into the app again, status bar in landscape again. So the whole view will be repositioned on every such situation. And there are more elements like navigation bar make a difference.

Anyway:
- this is not caused by this PR
- I will never do anything with it, even will not try.

opusforlife2 commented:
> this is not caused by this PR

Neve said it was.

> I will never do anything with it, even will not try.

It's not worth fixing anyway. I just noticed it when I was comparing padding between blackbox's PR and the latest unified debug, so I asked.

avently commented:
@opusforlife2 
> Neve said it was.

I never said that you said:)

I said it just for future reviewers of the PR. So it means that this PR is fine.

>It's not worth fixing anyway.

Actually I thought you'll open an issue but you agree with me. Bravo! It was unexpected :)

MD77MD commented:
🤣🤣🤣 both of you guys are just hilarious especially avently

avently commented:
Added a commit related to Android 11 #4277. There is still a problem with overlapping thin navigation bar but I didn't fix it because it is already fixed but unreleased in #4255

avently commented:
I have something interesting today. After I initially made this https://github.com/TeamNewPipe/NewPipe/blob/dev/app/src/main/java/org/schabi/newpipe/player/VideoPlayerImpl.java#L1492
I thought it's too complex thing to maintain but I didn't find another way. Make once right and you never need to touch this `black box`. But from time-to-time we see new issues related to this code. After I saw how much needed to be done in order to make my code working here #4255 I couldn't sit back and relax and started to search an alternative.

So I found a new (old) way of how player positioning should work. It doesn't need to have all these manual calculations. It just works (almost identical to 0.19.8).

I checked in portrait/landscape, vertical/non-verical videos, with/without cutout, on 4.4 (emulator), 9, 11 (emulator), on Android TV (only a little bit), in multiWindow mode. Maybe some bugs still present on real devices, can't say. 

It's not a `small fix` so it shouldn't be in this PR but since I started to work in that branch I left it as is. I can create a new PR with just this one commit if you want but this commit uses the changes from this PR anyway.

As I said, it works like in 0.19.8 but I don't like how shadow covers the player. Shadow should be reimagined too but not from me or not now 100%. For non-cutout the shadow works ok, but for cutouts it isn't as good as I want it to be.

Let me know if some bugs are present.

@opusforlife2 @TobiGr @Stypox @blackbox87 @wb9688 @B0pol 
@ZeroPucks

[positioning.zip](https://github.com/TeamNewPipe/NewPipe/files/5224611/positioning.zip)

opusforlife2 commented:
Sorry, it's not exactly clear what you've changed from a user perspective. Can you clarify what to test?

Stypox commented:
@avently with that positioning APK on Huawei P9 lite android 7.0, the controls go behind the system ui (hardware buttons and top dropdown menu indicator) instead of being constrained by them

avently commented:
@opusforlife2 we have a video view that should be located under status bar and navigation bar in landscape, under status bar in multiwindow mode, away of statusbar in portrait.
We have controls (top and bottom menu buttons/position slider, fullscreen button that located inside the player view). And the controls should be accessible in any orientation, device size, device type, with or without cutout, etc. So the controls shouldn't be overlapped by other visual elements.

So if you see some button/control is inaccessible or located at a wrong position in the player view it means you find a bug.

avently commented:
@Stypox but this device doesn't have `hardware buttons`, only virtual buttons.

Can you show a screenshot?

avently commented:
@Stypox I installed Android 7.0 emulator and everything works fine, how to reproduce the issue you reported?

opusforlife2 commented:
Alright. I have a simple display. Hardware buttons and no cutouts/notches. Here's my testing comparing base unified and positioning apks:

- Positioning: when you tap to show player controls and rotate to landscape, you can see that after the status bar hides itself, the shadow doesn't extend to cover that area, leaving a weird bright rectangle until the controls fade out.
- Unified: the shadow correctly extends to that area.

This happens with auto-rotate both off and on.

That's it! That's the only bug I found.

People with cutouts/notches/navigation bars may have other issues, though, so I suggest trying this experiment in a separate Issue, and if it works out, a PR later.

Stypox commented:
> but this device doesn't have hardware buttons, only virtual buttons.

Yeah, sorry, I meant those buttons. Anyway, I can't reproduce. Either I tested the wrong app (though I am pretty sure I tapped on "Open" after having installed the apk you provided), or who knows. I didn't do anything strange to create the situation I described above, I just tapped on a random video in trending and tapped on the fullscreen button.
@avently

avently commented:
@opusforlife2 
> when you tap to show player controls and rotate to landscape, you can see that after the status bar hides itself, the shadow doesn't extend to cover that area, leaving a weird bright rectangle until the controls fade out.

> but I don't like how shadow covers the player. Shadow should be reimagined too but not from me or not now 100%

Fiiine. Made a perfect shadow. Shadow of all shadows! 
If no problem will be found, the PR can be merged.

@Stypox 
>Anyway, I can't reproduce

> or who knows. 

For me with every configuration and on any devices I test the app works perfectly.
In this apk: 
- perfect shadow. No one has this shadow but you.

[position-and-shadow.zip](https://github.com/TeamNewPipe/NewPipe/files/5226532/position-and-shadow.zip)

opusforlife2 commented:
> Fiiine. Made a perfect shadow. Shadow of all shadows!

The shadow is indeed perfect!

New bug, though. Lock phone in portrait orientation. Play a video. Tap on the fullscreen button: you can see the player controls jump down.

avently commented:
@opusforlife2 can't reproduce. Maybe you can make a screenrecord? What android version?

opusforlife2 commented:
Android 9. Give me a few minutes for the screen record.

opusforlife2 commented:
Here you go: [ScreenRecord.mp4.gz](https://github.com/TeamNewPipe/NewPipe/files/5226770/ScreenRecord.mp4.gz)

avently commented:
> Here you go

Thanks. Looks like when I re-adding the player after orientation change, Android is not able to adjust the player's padding so fast and that's why you see unadjusted paddings first and after tens milliseconds you'll see adjusted paddings. I tried to force Android to adjust the padding even before adding the player to view hierarchy but without success. 

In this apk after moving from portrait to landscape the controls will be hidden automatically. I don't know if it helps in your situation or not, please, check.
[position2.zip](https://github.com/TeamNewPipe/NewPipe/files/5227317/position2.zip)

opusforlife2 commented:
@avently Thanks! That looks great! I think hiding controls after rotation is better than the previous behaviour in terms of usability. And you got rid of the bug as a bonus!

Would it be trivial to implement this the other way around also? Hide controls when you tap fullscreen button to go from landscape to portrait?

opusforlife2 commented:
What's that grey bar at the bottom, though? Immediately after rotation to landscape, you can see a grey bar over the seekbar which disappears quickly. It is visible in my screen record.

opusforlife2 commented:
Also, the time taken for the rotation process to complete has been cut down from around 1 second in the unified apk to around 0.5 second in the position2 apk!

ALSO, seeking happens much faster now! @blackbox87 check it out!

avently commented:
@opusforlife2 
>Would it be trivial to implement this the other way around also? Hide controls when you tap fullscreen button to go from landscape to portrait?

Just to uncomment one line of code. I didn't do this for landscape to portrait because often when I switching to portrait I want to somehow interact with controls. For example, to tap on more options button or on pause button. So I prefer to show controls in this case.

>What's that grey bar at the bottom, though?

I saw this color on SystemUI elements like navigation bar, status bar. I don't know why in your case you see it since you don't have navigation bar at all. 


>ALSO, seeking happens much faster now

The changes I made are unrelated to seeking. Nothing near to it. If you're saying about the last commit, of course. If you're saying about the whole PR...I still don't see how it can make seeking faster.:)

> Also, the time taken for the rotation process to complete has been cut down

This can be true, yes.

opusforlife2 commented:
¯\\\_(ツ)\_/¯

Blackbox was saying something about heavy calculations being done in the unified player. Maybe those were reduced by your changes?

avently commented:
Calculation is complex for developer not for hardware in the case of old code. So, no, I don't think I changed something related to performance.

ZeroPucks commented:
> 
> 
> > Here you go
> 
> Thanks. Looks like when I re-adding the player after orientation change, Android is not able to adjust the player's padding so fast and that's why you see unadjusted paddings first and after tens milliseconds you'll see adjusted paddings. I tried to force Android to adjust the padding even before adding the player to view hierarchy but without success.
> 
> In this apk after moving from portrait to landscape the controls will be hidden automatically. I don't know if it helps in your situation or not, please, check.
> [position2.zip](https://github.com/TeamNewPipe/NewPipe/files/5227317/position2.zip)

Tested this APK on Android 11, looks good. Progress bar is still right on top of the gesture navigation pill, but it is much easier to manipulate and drag-to-seek without accidentally swiping to change apps.

Changing from portrait to landscape using the fullscreen button is pretty fast, no complaints there myself.

Only other observation I have is that the status bar overlaps with the top UI elements like the video title in the top left, as seen in the screenshot:
![Screenshot_20200915-184540_1](https://user-images.githubusercontent.com/46003124/93272764-4692f000-f784-11ea-9924-f1f3f69e8918.png)

avently commented:
@ZeroPucks thank you for testing. Are you sure that you ran correct apk from `position2.zip` file? Could you re-check? Because my main emulator for testing of this PR was Android 11 and here is what I see using the apk:
![Screenshot_1600211476](https://user-images.githubusercontent.com/7953703/93274223-5f61cc80-f7a9-11ea-9a2e-e9831d1db451.png)


I don't know maybe there is such difference between emulator and real device but it's strange. If you use correct apk I probably need to find a real device with Android 11.

ZeroPucks commented:
@avently I think I found the issue. When opening a video from anywhere like the Trending tab or from search results, the status bar and seekbar padding are perfect, just like your screenshot. But opening a video from the Subscriptions tab or a channel page that is subscribed too results in less padding, just like my screenshot.

blackbox87 commented:
@avently It looks good to me when using Android 9, although I'd need to install an Android 10 ROM again to see if it handles things correctly when you configure gesture navigation to hide the pill.

avently commented:
@ZeroPucks 
> But opening a video from the Subscriptions tab or a subscribed channel page results in less padding

Yep, able to reproduce. Thank you for such interesting finding. Will see what I can do

blackbox87 commented:
@avently When using LineageOS 17.1 and the option to hide the pill NewPipe will draw over the gesture area. This happens because the height of the navbar becomes 0, yet it still reacts to touches in that area.

![Screenshot_20200915-175239_Settings](https://user-images.githubusercontent.com/39933433/93279929-6516dc80-f7c0-11ea-95ad-ce1d12fd4641.png)
![Screenshot_20200915-175107_NewPipe_SmallFixes2](https://user-images.githubusercontent.com/39933433/93279939-69db9080-f7c0-11ea-8bec-d6e28440ee26.png)
![Screenshot_20200915-175239_Settings](https://user-images.githubusercontent.com/39933433/93279945-6c3dea80-f7c0-11ea-8fc1-f840e247d297.png)

You could do what I do and check if gesture navigation is enabled and the navbar height is 0. If it is then apply some extra padding depending on the current orientation.

avently commented:
@ZeroPucks fixed the issue you reported. There were problems when multiple fragments tried to receive padding (insets) from the system (only one listener is allowed). Now should be fine.

[position3.zip](https://github.com/TeamNewPipe/NewPipe/files/5229159/position3.zip)

@blackbox87 
> If it is then apply some extra padding depending on the current orientation.

there are no problems with it right now.

blackbox87 commented:
> there are no problems with it right now.

It's drawing over the gesture area, so sometimes it might react to you pressing on the seekbar and sometimes it might think it's a gesture. This issue isn't unique to LineageOS either as a few other ROMs have the same option e.g. Pixel Experience.

MD77MD commented:
@avently seeing that you are solving little issues here, I'm not sure if i sould open new issues or not for this bug.

anyway, if you open the video (link below) in landscape mode, the "video resolution list" is not scrollable anymore like in 19.8... this means you can not see all available resolutions... although they are showing in portrait mode.

https://www.youtube.com/watch?v=jbi9ScRgitI&t=35

p:s. using latest unified debug apk

opusforlife2 commented:
@MD77MD Open a new issue for your bug anyway. It's so that it can be documented and referred to later when needed.

avently commented:
@MD77MD try the latest apk in this PR, there is no such problem here. Let me know if you see the problem in the latest apk too. And don't listen opusforlife2 (he like to create issues just because he likes it)

avently commented:
@blackbox87 since you're referring to custom feature of your OS I don't want to do something with it. It's not my use-case. You can make a PR if you want but note that it shouldn't be default behavior because 99% of users with hardware buttons don't use custom feature of your rom. So it can be selected from settings but not default.

blackbox87 commented:
> @blackbox87 since you're referring to custom feature of your OS I don't want to do something with it. It's not my use-case.

> but note that it shouldn't be default behavior because 99% of users with hardware buttons don't use custom feature of your rom

If the option was unique to a random custom ROM then I'd agree with you, but the option exists in LineageOS, ArrowOS, Resurrection Remix, Pixel Experience, AICP and many more. And NewPipe is used by a lot of people who uses these ROMs because you don't have to have GAPPS installed, so I'm not sure where you've got that 99% figure from as that's likely nowhere near accurate.

The issue might not impact your use cases or even my own, but there are people out there who are using the option. So why not fix it now so that someone else doesn't have to report it and then fix it with a solution similar to my own? It just doesn't make sense to me. Having said that your approach to fixing issues seems different than my own since I often test my changes against every relevant OS option that I can think of, but your approach seems to be that if an issue doesn't impact your personal use cases then people can deal with it or fix the issue themselves. And that's exactly why I had to submit my own PR to handle cutouts in the first place. But the difference between your PR and my own is that I tested my changes against a few ROMs on real hardware, so I covered more use cases and have fixes in place for the known issues.

> You can make a PR if you want

I won't submit another PR since I've already demonstrated the problem and shared a solution. Plus my existing PR already fixes the issue.

avently commented:
> And NewPipe is used by a lot of people who uses these ROMs because you don't have to have GAPPS installed

I always use custom ROM. Lineage is my favorite. It doesn't mean I use the feature you are talking about. I just don't need it. When Android 10 was released, users can use system-wide thin navigation bar which is (probably) better aproach since it supported by every app natively (because of automatic insets). Instead of this custom thing from custom ROMs. This is why I said about 99% people that will not use the custom gesture navigation.

> The issue might not impact you or me, but there are people out there who are using the option

> but your approach seems to be that if an issue doesn't impact your personal use cases then people can deal with it or fix the issue themselves

So what? There are 626 issues currently open and I don't see that you are trying to fix any of them except the bug that impacts your current use-case, isn't it? You're saying that (based on your words) I should make code, spend my time for free without any interest in it based on someone's use-cases? It's insane. It's not my app and I don't care about problems that exist in software (except the ones that caused by me or the ones that needs to be made for me by me). Hello, it's free and open source software that made by people who want to do something. Trying to blame such people (us) that they don't want to do what you want to receive is a nonsense.

For some reason even you (which is strange, because you are a dev) don't want to understand that other people have their lifes and they should pay money for things in their lifes (like a house, cars, family). And all of that takes some time.

> but your approach seems to be that if an issue doesn't impact your personal use cases then people can deal with it or fix the issue themselves

It's not actually true because I did only 20-30% of things I need, other things were made on response to requests from the people here. That's because I think that the requests were clever/useful enough for me to spend time on it. You can review the whole PR and to ensure that I'm saying the truth #2907. 

> But the difference between your PR and my own is that I tested my changes against a few ROMs on real hardware, so I covered more use cases and have fixes in place for the known issues.

I said this:
>I checked in portrait/landscape, vertical/non-verical videos, with/without cutout, on 4.4 (emulator), 9, 11 (emulator), on Android TV (only a little bit), in multiWindow mode. Maybe some bugs still present on real devices, can't say.

So what's the difference you're saying about?:)

> Hardware buttons shouldn't be a factor because if you've got gesture navigation enabled then it'll activate a software navbar. And that navbar will always have a height that's greater than 0 until you enable the option to hide the pill.

The problem with your approach that you never know that you need to adjust padding. So some device don't have nor custom gesture control, nor virtual buttons, they have only hardware buttons. And for all of them you'll adjust the padding which will make the app looking different on devices. Which will create a possibility for bugs in a feature because it will be harder to check and to maintain. But looks like you don't care which is not the case for me. I want to make something that will work for everyone. Yes, sometimes I don't know how to make it possible but I'm trying. And this PR about that: I made a universal solution for positioning the player that makes your PR unneeded and works on every device without dirty hacks.

Your words: https://github.com/TeamNewPipe/NewPipe/pull/4255#issuecomment-690615580
> I think it is a problem since it worked perfectly before the unified player was added. That's the only reason I'm still trying to fix this.
At the moment I can't say that I'm a fan of the unified player as it's introduced more issues than it solves.

In this PR I tried to make the situation better by removing possible place for bugs but you still don't like it. 

I already said that you did an excelent job and I really thinks so. I just found that we don't need to do so much things like we do. The situation is much simpler and doesn't require from us manual calculations. 

>I won't submit another PR since I've already demonstrated the problem and shared a solution. Plus my existing PR already fixes the issue.

It's your choice. You can submit the PR or not, it's not what I would ask from you because as I said it's free and open source software, you can do what you want. And even if you show a solution, it's your code and I don't want to support it in case of bugs that could be found in it in the future. Your code = your responsibility. So if you want you can submit it or not.

> Plus my existing PR already fixes the issue

And you really think that my old code + code from your PR should be used instead of this new universal & easy code from this PR? If so, why do you like this custom hell that was started from my initial wrong way of doing things?

blackbox87 commented:
> When Android 10 was released, users can use system-wide thin navigation bar which is (probably) better aproach since it supported by every app natively (because of automatic insets). Instead of this custom thing from custom ROMs. This is why I said about 99% people that will not use the custom gesture navigation.

LineageOS 17.1 has the same gesture navigation settings as AOSP, but it also includes this additional setting. And that setting has been ported to dozens of other custom ROMs. So no, you can't guarantee that 99% of people won't use it. But even if 99% of people didn't use it then you're basically telling the 1% that you don't care about them, despite a fix existing with no negative side effects.

> So what? There are 626 issues currently open and I don't see that you are trying to fix any of them except the bug that impacts your current use-case, isn't it?

I've previously contibuted code to NewPipe Legacy, despite me not using it. And I only pushed a PR that fixes cutouts because it previously worked almost perfectly in v0.19.8 and you seemed unwilling to fix it until I did.

> You're saying that (based on your words) I should make code, spend my time for free without any interest in it based on someone's use-cases?

Well, every bit of code that I push I test across multiple devices. And I do that becasuse unlike iOS you need to check that things behave correctly on a variety of devices with different flavours of Android.

But what I'm saying is that if someone says "hey, there's an issue and here's a solution" then you should listen to them and either use their solution or come up with your own to fix the problem. Like yeah, it doesn't impact you personally, but it doesn't mean that the issue doesn't exist and can be ignored. It exists and there will be people who use the setting, whether you like it or not.

I don't have all of the time in the world to do whatever I want either, but when you orignally reported the issue with my commit did I ever say that it's not a problem because I couldn't reproduce your issue? No. I investigated it further, found the problem, tested the fix and then shared it. The same goes for my PR that fixes the rotation issues.

> So what's the difference you're saying about?:)

I tested with every possible relevant option on LineageOS 14.1, LineageOS 16, LineageOS 17.1 & Pixel Experience across 3 real devices (2 phones and 1 tablet) and a couple of emulators. And the difference is that I found an issue that I could reproduce across multiple ROMs, so I fixed it in my PR. But based off of what you've said it's clear that you don't want to fix it because you don't think the options important enough and it doesn't impact your personal use cases.

> The problem with your approach that you never know that you need to adjust padding. So some device don't have nor custom gesture control, nor virtual buttons, they have only hardware buttons.

Think about that. My patch only applies when the navbar mode is 2 and the height is 0. So how is the navbar mode going to get to 2 if the ROM doesn't support gesture controls due to the hardware navbar? And even if a custom ROM supported it then when you enable gesture controls the navbar height becomes the size of the pills mini navbar and not the devices actual navbar. And I know that because I've tested it already.

> And for all of them you'll adjust the padding which will make the app looking different on devices

If you apply 16dp in landscape and 4dp in portait then it'll look correct on various devices as it scales up or down depending on your screen size and density. And even if the extra padding isn't perfect it's still an improvement as the currently solution guarantees that NewPipe's seekbar is drawn directly over the hidden gesture bar.

> In this PR I tried to make the situation better by removing possible place for bugs but you still don't like it.
I already said that you did an excelent job and I really thinks so. I just found that we don't need to do so much things like we do. The situation is much simpler and doesn't require from us manual calculations.

To quote myself "[I'm happy with whatever as long as this gets fixed](https://github.com/TeamNewPipe/NewPipe/pull/4154#issuecomment-687993427)". I just don't agree with ignoring problems, which my commit fixes.

> And you really think that my old code + code from your PR should be used instead of this new universal & easy code from this PR? If so, why do you like this custom hell that was started from my initial wrong way of doing things?

There's no question that your method is cleaner. I mean that's to be expected too since you're a lot more familiar with the code than me. But you can't deny that my version currently supports more use cases.

avently commented:
> despite a fix existing with no negative side effects

It is what you think but the reallity is not so easy. Someone may find a bug in this code or Android will change something in the future. You can't be sure. Even if this code will work ideal forever it's your code, you can make a PR with it or not, I don't want to push your code.

> And I only pushed a PR that fixes cutouts because it previously worked almost perfectly in v0.19.8 and you seemed unwilling to fix it until I did.

But it doesn't mean you're fixing all those 626 issues NewPipe has. And I shouldn't do it too. And non of your arguments can make me thinking that I should.

> But what I'm saying is that if someone says "hey, there's an issue and here's a solution" then you should listen to them and either use their solution or come up with your own to fix the problem

Disagree. Someone gives you a code saying that it fixes some issues you don't even care/know about and you're saying that you will automatically push it to the master? Oh, that's bad. Don't think it's a good situation.

> But based off of what you've said it's clear that you don't want to fix it because you don't think the options important enough and it doesn't impact your personal use cases.

I don't want to fix it because there is nothing to fix. Everything works as expected. If you have different vision, you can make a PR. That's it.

> Think about that. My patch only applies when the navbar mode is 2 and the height is 0.

Oh, wait. You know more information for YOUR code than me, hm, isn't it mean that you probably the best person who need to make a PR?:)

>  I just don't agree with ignoring problems, which my commit fixes.

Same for me. That's why you can find multiple PRs here when I fixed some problems with my code.

> But you can't deny that my version currently supports more use cases

The code you're asking to include doesn't rely on your changes, isn't it? So you can just paste it inside my approach. If it's not true than it means you're asking about adapting YOUR code to my code which will cost me some hours/days to figure out how to make it working universally.

What stops you from making a PR based on changes from this PR? If agree too that the method used here is cleaner.

blackbox87 commented:
> It is what you think but the reallity is not so easy. Someone may find a bug in this code or Android will change something in the future. You can't be sure. Even if this code will work ideal forever it's your code, you can make a PR with it or not, I don't want to push your code.

The odds of navbar mode 2 and the height being 0 not being an issue are extremely slim, but should it ever be an issue then the code could be changed to suite.

> Disagree. Someone gives you a code saying that it fixes some issues you don't even care/know about and you're saying that you will automatically push it to the master? Oh, that's bad. Don't think it's a good situation.

But the code has already been tested and you even said that it "[all works perfectly](https://github.com/TeamNewPipe/NewPipe/pull/4255#issuecomment-691706783)" 🤔

> I don't want to fix it because there is nothing to fix. Everything works as expected. If you have different vision, you can make a PR. That's it.

See, that's just being ignorant. I've shown the issue, explained the issue and provided a fix.

> Oh, wait. You know more information for YOUR code than me, hm, isn't it mean that you probably the best person who need to make a PR?:)

Like the PR that already exists and works perfectly? Is it as pretty as your version? No. But it works.

> So you can just paste it inside my approach

With very minor changes, yes. But I'm not about to make another PR when my existing PR fixes the problem.

It's pointless for me to say much more since you don't think a problem is a problem, even though conflicting input areas is a problem.

avently commented:
Made a fix for #4143. @nyanpasu64, take a look

[touchable.zip](https://github.com/TeamNewPipe/NewPipe/files/5234837/touchable.zip)

MD77MD commented:
i know you guys mean well but please let's just stick to constructive conversations... if anyone have a good idea share it and let everyone build on it "positively" and nothing else please🙏...at the end everyone have a limited time for newPipe so...
anyway everyone's effort is appreciated no matter how small😉

MD77MD commented:
@avently this problem https://github.com/TeamNewPipe/NewPipe/pull/4272#issuecomment-693286809 is fixed.. anyway there is a small shift in the list position when you start scrolling... however if you think it will take too much time to fix just leave it be

avently commented:
> anyway there is a small shift in the list position when you start scrolling

What do you mean? can you show a screenrecord?

MD77MD commented:
I mean 
1. open the video like in my comment in landscape mode
2. open the list
3. now if you scroll the list up and down the whole list will shift to that left (just a little)

anyway if this doesn't happen with you then just ignore my comment... its not important

nyanpasu64 commented:
> Made a fix for #4143. @nyanpasu64, take a look
> 
> [touchable.zip](https://github.com/TeamNewPipe/NewPipe/files/5234837/touchable.zip)
> 

Not fixed. https://youtu.be/zQWniF90J_8

avently commented:
@nyanpasu64 
> Not fixed

I though the problem you're saying about was untouchable area where bottom controls are located. So now that area works properly. 
In your case you swipe the view down, than up, releasing a finger, and quickly swipe down again when the player is not stabilized yet. I don't think someone will ever do this. So you probably expect in this situation to get the player hiding down but it will not work like this. In the latest commit I made an adjustment that in your situation force the player expanded instead of shaking, no movements will be made until your release the finder and repeat. If you don't like the result I have nothing to suggest.

In this apk:
- made an exit from fullscreen in case of player error in a correct way
- set alpha for mini player when bottom sheet stabilizes in expanded or collapsed state (not only in dragging state like before). Prevents rare situation when alpha is incorrect

[final.zip](https://github.com/TeamNewPipe/NewPipe/files/5240633/final.zip)

Ready for review. Today I worked on player performance and the result will be in the next PR which is based on this PR.

nyanpasu64 commented:
> I though the problem you're saying about was untouchable area where bottom controls are located. So now that area works properly.

I forgot which issue number you're talking about :(

> In the latest commit I made an adjustment that in your situation force the player expanded instead of shaking, no movements will be made until your release the finder and repeat. If you don't like the result I have nothing to suggest.

I tried this build, and the new behavior seems acceptable to me. Thanks!

opusforlife2 commented:
> In the latest commit I made an adjustment that in your situation force the player expanded instead of shaking, no movements will be made until your release the finder and repeat.

Perfect! Thanks, @avently!

Stypox commented on app/src/main/java/org/schabi/newpipe/views/CustomCollapsingToolbarLayout.java:
```diff
@@ -0,0 +1,36 @@
+
+    /**
+     * CollapsingToolbarLayout overrides our logic with fitsSystemWindows and ruins the layout.
+     * Override Google's method
```
Could you better explain in which way this overrides a method? At first I interpreted it as a Java-language override, but it is not and you meant override in the sense of overwrite, so please change this to "Override Google's behaviour"

avently replied:
Done

Stypox reviewed (changes requested):
Are you sure you can remove `showControls(DEFAULT_CONTROLS_DURATION)` without any issue on Android TVs, where the user can be slow at moving dpad to other buttons?

avently commented:
@Stypox
> Are you sure you can remove showControls(DEFAULT_CONTROLS_DURATION) without any issue on Android TVs, where the user can be slow at moving dpad to other buttons? 

I doesn't matter how slow the user is because the controls will not be hidden when popup is visible (it's the case for quality selector and caption selector), see https://github.com/TeamNewPipe/NewPipe/pull/4272/files#diff-61ebd1c28e394d4dbe7ffd9c9402e610L806
For playback speed the controls is useless. So in this three places I can skip showing the controls.

avently commented on app/src/main/java/org/schabi/newpipe/player/VideoPlayer.java:
```diff
@@ -359,8 +363,8 @@ private void buildCaptionMenu(final List<String> availableLanguages) {
             if (userPreferredLanguage != null && (captionLanguage.equals(userPreferredLanguage)
                     || searchForAutogenerated && captionLanguage.startsWith(userPreferredLanguage)
                     || userPreferredLanguage.contains("(") && captionLanguage.startsWith(
```
I didn't write this mess there is only a spacing change from IDEA. Anyway I added parenthesis:)

avently commented:
@B0pol please, don't edit my posts. If you want to link the issue for closing you can link it with Github button like @opusforlife2 do.

In this apk: 
- caption button has variable width and it works in both players
- system UI will be hidden when popup closes (after quality change or after caption selection, etc)

[final2.zip](https://github.com/TeamNewPipe/NewPipe/files/5250535/final2.zip)

opusforlife2 commented:
Werks. 👍

MD77MD commented:
>system UI will be hidden when popup closes (after quality change or after caption selection, etc)

works great on both normal and pop-up mode. 

thank you @avently for this, also is possible to make the same when double taping for pause.

opusforlife2 commented:
There is no 'double tap to pause' gesture. It's just that one tap reveals the controls and a second tap on the pause button pauses, so if you double tap in the pause icon area you'll actually pause. But it's not a coded in gesture.

MD77MD commented:
i meant for the player ui to stay hidden after a "double tap to pause"

opusforlife2 commented:
What's the utility of hiding the controls upon pausing? Generally, a user would want to pause when they want to use the player UI, like the Share button, or change of aspect ratio or subtitles.

MD77MD commented:
@opusforlife2

I did consider this. but think about it, we usually do them without pausing the video, except in rare cases. 

however, if we consider the cleaner UI we can git whenever a quick pause and play action is needed such as for reading a chart, comparing two pics or even taking a screenshot, etc.

all of these actions can only be done if the video is paused giving them a priority in  considering keeping the controls out of there way... 
 
side note: it also unifies the double tap to seek/pause gestures behavior related to player controls showing..

opusforlife2 commented:
We probably need more user feedback on this. Personally, I like that the controls stay on screen indefinitely after pause, because I can immediately tap on it again to resume the video. Like if I pause the video because I need to talk to someone for a few seconds, then resume watching.

avently commented:
Agree with @opusforlife2

MD77MD commented:
>Personally, I like that the controls stay on screen indefinitely after pause, because I can immediately tap on it again to resume the video. 

You can, just double tap to resume... actually this IS another scenario for hiding player controls i usually encounter... anyhow, what I would like you guys to  do is to keep what i suggested in mind... will give those scenarios a chance for some days, maybe you may like it then.. 😄🙌

also we can open another thread to get others thoughts like what you mentioned

opusforlife2 commented:
Sure, open a new thread for this, @MD77MD. Just a poll/discussion regarding showing/hiding the player UI after pausing. Sweet and simple.

MD77MD commented:
Yes sir👨🏻‍✈️.....😁

TobiGr commented:
There seem to be problems when changing aspect ratio of vertical videos when using the embedded player (tested with KitKat emulator):
![kitkat_vertical_video](https://user-images.githubusercontent.com/17365767/94274407-f9faa200-ff45-11ea-9162-b02af51ba402.gif)

TobiGr commented:
@MD77MD can you please confirm that my emulator did not do something stupid again?

MD77MD commented:
@avently  I would like to know if your aware of another brightness bug... the brightness would sometime reseat to zero when you try to change it.  have you experienced it and does this PR fix it?

@TobiGr  off course but until tomorrow because it's with my little brother

avently commented:
@TobiGr
> There seem to be problems when changing aspect ratio of vertical videos when using the embedded player (tested with KitKat emulator):

Wow, it's unexpected. Same thing happens on `dev` branch, so it's totally unrelated to this PR.

Looks like on KitKat setScaleX/Y works in different way than on other Android versions, this is where it is applied:
https://github.com/TeamNewPipe/NewPipe/blob/dev/app/src/main/java/org/schabi/newpipe/views/ExpandableSurfaceView.java#L75

So when you pass 1 for scaling here https://github.com/TeamNewPipe/NewPipe/blob/dev/app/src/main/java/org/schabi/newpipe/views/ExpandableSurfaceView.java#L75 and here https://github.com/TeamNewPipe/NewPipe/blob/dev/app/src/main/java/org/schabi/newpipe/views/ExpandableSurfaceView.java#L76 
everything works ok (without correct dimensions of course). Maybe it's an emulator bug? I don't believe that such important thing changed so much from KitKat to newer Androids.

Also there is a problem related to portrait-to-fullscreen switch via button: you can see that the bottom side of the player is not fully expanded, this can be fixed by replacing this code: https://github.com/TeamNewPipe/NewPipe/blob/dev/app/src/main/java/org/schabi/newpipe/views/ExpandableSurfaceView.java#L76   with this:

```java
new Handler().postDelayed(this::addVideoPlayerView, 0);
```
Looks like without delay KitKat returns a wrong value via `activity.getWindow().getDecorView().getHeight()`. 
I also want to listen  @MD77MD feedback on these two issues. Maybe it's only the emulator problem. @MD77MD  do you have two of these problems on your device?

@MD77MD 
> the brightness would sometime reseat to zero when you try to change it. have you experienced it and does this PR fix it?

I don't know about that. Could you describe the steps I need to do in order to reproduce the issue?

blackbox87 commented:
> do you have two of these problems on your device?

I'm not @MD77MD, but I can tell you that it also happens on Android 4.4.4 via Genymotion too.

If you use NewPipe v0.19.8 and toggle between fit, fill & zoom while watching a vertical video in fullscreen then the button doesn't do anything. And even if you change the aspect ratio while in landscape and then switch back to portait it'll ignore the setting.

So I guess you could just copy what v0.19.8 does and ignore the aspect ratio setting if you're playing a vertical video?

> Also there is a problem related to portrait-to-fullscreen switch via button: you can see that the bottom side of the player is not fully expanded

Confirmed. And here's what that looks like.

![emu_1](https://user-images.githubusercontent.com/39933433/94326398-900ce780-ff9b-11ea-82a8-ebf2341ae9aa.png)

avently commented:
@blackbox87 I use Genymotion too in case of KitKat.

avently commented:
@MD77MD 
>off course but until tomorrow because it's with my little brother

Any news?

MD77MD commented:
@avently I'm sorry but my family is still traveling... anyway i have another device with a broken screen, I'll try using that... although it pretty frustrating to do anything on it.

MD77MD commented:
ok so @avently I tried this apk https://github.com/TeamNewPipe/NewPipe/pull/4272#issuecomment-695350659 and got the same result as https://github.com/TeamNewPipe/NewPipe/pull/4272#issuecomment-698937985 ... although the video would go back playing normal if you return the aspect ratio to  fit... moreover, normal landscape videos don't suffer from this problem, on the other hand, probably because all three aspect modes have no effect what's so ever... perhaps we can test this if there is a landscape video with a weird aspect ratio.

> also want to listen @MD77MD feedback on these two issues.

do you mean by the second issue the brightness issue?

avently commented:
> do you mean by the second issue the brightness issue?

This one:
https://github.com/TeamNewPipe/NewPipe/pull/4272#issuecomment-699252908

Also let me know how to reproduce the brightness issue.

And thank you for the testing. Looks like the only thing I can do is to disable ZOOM mode for KitKat devices...  so this mode will act like FIT mode.

MD77MD commented:
>I don't know about that. Could you describe the steps I need to do in order to reproduce the issue?

@avently I have been trying for the past week to catch the exact steps to reproduce this bug but failed. however this is the closest I got:

1. open a video and set brightness lets say 40 or 70% ... close or pause the video
2. switch to another app 
3. go back to newPipe
4. open any video in landscape mode... the video will start playing using the same last brightness you used (everything is fine until now)
5. the bug: if you would try increase the brightness the screen brightness would suddenly drop and the slider would starts to increase from 0% even if the last value was greater.

expected behavior: when changing brightness... it would start from the same exact point used on the last video.

avently commented:
@MD77MD

> the bug: if you would try increase the brightness the screen brightness would suddenly drop and the slider would starts to increase from 0% even if the last value was greater. 

This can happen before this PR and I fixed such situation here. Are you sure you see such behavior in `final2.zip`?

This

avently commented:
In this apk:
- fixed #4322 
- fixed https://github.com/TeamNewPipe/NewPipe/pull/4272#issuecomment-698937985
- fixed (KitKat specific probably) this bug -> `Also there is a problem related to portrait-to-fullscreen switch via button: you can see that the bottom side of the player is not fully expanded`
- made an enhancement https://github.com/TeamNewPipe/NewPipe/pull/4254#issuecomment-699537690
- fixed rotation via fullscreen button when a device in landscape in fullscreen with locked global orientation. So now in this case the device will rotate to portrait after press on fullscreen button (and back after another press). The fix is just one line, lol

[final3.zip](https://github.com/TeamNewPipe/NewPipe/files/5287582/final3.zip)

MD77MD commented:
tried final3 https://github.com/TeamNewPipe/NewPipe/pull/4272#issuecomment-699570617 on android 4.4 and does solve the mentioned problems.

avently commented:
Same as https://github.com/TeamNewPipe/NewPipe/pull/4272#issuecomment-699570617 but with some changes:

- made refactoring and included one suggestion to rotate the screen when in landscape & fullscreen on vertical video. So now the app will be rotated to portrait in this case after clicking on fullscreen button.

[final4.zip](https://github.com/TeamNewPipe/NewPipe/files/5287697/final4.zip)

opusforlife2 commented:
I'm reporting a small visual bug here which is present in the unified debug, since it probably has a small fix as well:

Play a playlist or some other queue, and open the queue from the main player. Swipe down in the middle. The queue will rapidly vibrate up and down.

TobiGr reviewed (commented):
Looks good to me! 

Thank you for your ongoing efforts :heart:

avently commented on app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java:
```diff
@@ -1927,21 +1946,29 @@ public void onFullscreenStateChanged(final boolean fullscreen) {
+            addVideoPlayerView();
+        } else {
+            // KitKat needs a delay before addVideoPlayerView call or it reports wrong height in
+            // activity.getWindow().getDecorView().getHeight()
```
It's a part of an explanation/comments. It probably should be written in JavaDoc style

TobiGr replied:
all good, i was a little sleepy while reviewing your changes this morning.

avently commented:
@opusforlife2 
>The queue will rapidly vibrate up and down.

Vibrate? What do you mean? Show a screenrecord, nothing unexpected happens for me

opusforlife2 commented:
Give me a minute.

opusforlife2 commented:
Wow. Just... wow. The video was large so I scaled it down using ffmpeg from 1080 to 720. Somehow the size went from 43 MiB to under 2 MiB. O.o


[ScreenRecord.mp4.gz](https://github.com/TeamNewPipe/NewPipe/files/5288206/ScreenRecord.mp4.gz)

TobiGr commented:
I was able to reproduce this, too. only happens, when the status bar is hidden. fir some reason, on a pixel 3a XL simulator with Android 11, the status and navigation bar are always displayed (but without content) when playing a video:
![Screenshot_1601207142](https://user-images.githubusercontent.com/17365767/94364189-db241900-00c7-11eb-94b4-834a5739aa18.png)

avently commented:
@opusforlife2 
> ScreenRecord.mp4.gz

Thanks, reproduced.

>  fir some reason, on a pixel 3a XL simulator with Android 11, the status and navigation bar are always displayed (but without content) when playing a video

I use Pixel 3 emulator with Android 11 and no such problem. Are you sure you use the latest code or apk? Or do I need to do something special to reproduce?

TobiGr commented:
> I use Pixel 3 emulator with Android 11 and no such problem. Are you sure you use the latest code or apk? Or do I need to do something special to reproduce?

Never mind. Some noob tested the wrong version...

avently commented:
@opusforlife2  
> ScreenRecord.mp4.gz

fixed
[final5.zip](https://github.com/TeamNewPipe/NewPipe/files/5288302/final5.zip)

opusforlife2 commented:
> fixed
final5.zip

Confirmed!

TobiGr reviewed (approved):
Thank you, once again!
